Add password strategy compatible with simple-ember-auth : refactor branch

alexjp opened the pull request:
Is the branch refactor the correct branch? 

Rewrote changes from scratch, what is currently working :
- added registration by username ( email still works as before, all tests pass )
- when registering by username, does not send email and auto confirms
- allow to create session by username, sends the reply that simple ember auth expects

I think this patch is better and cleaner, although, the function names/if's could be better.

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/manukall/phoenix_token_auth/21)

<!-- Reviewable:end -->

Commit:
added support for creating session for simple ember auth

Commit:
add support for using username

Commit:
fix tests using new Authenticator.authenticate_email

Commit:
added registration by username and auto confirmation

manukall commented:
thanks! really happy you're contributing.

i'm gonna add a few minor comments to the files, but i like the general concept! :+1: :+1: :+1:

manukall commented on lib/phoenix_token_auth/authenticator.ex:
```diff
@@ -11,8 +11,16 @@ Returns:
     user = UserHelper.find_by_email(email)
+    authenticate(user, password)
+  end
+  def authenticate_username(username, password) do
```
please rename to `authenticate_by_email` / `authenticate_by_username`

manukall commented on lib/phoenix_token_auth/controllers/sessions_controller.ex:
```diff
@@ -15,8 +15,15 @@ defmodule PhoenixTokenAuth.Controllers.Sessions do
   Responds with status 200 and {token: token} if credentials were correct.
   Responds with status 401 and {errors: error_message} otherwise.
   """
+  def create(conn, %{"username" => username, "password" => password, "grant_type" => "password"}) do
```
please remove the `grant_type` param. it's not really relevant and without it, people not using ember can use this as well.

manukall commented on lib/phoenix_token_auth/controllers/users_controller.ex:
```diff
@@ -22,13 +22,19 @@ defmodule PhoenixTokenAuth.Controllers.Users do
-        Mailer.send_welcome_email(user, confirmation_token)
+        unless params["username"] do
+          Mailer.send_welcome_email(user, confirmation_token)
+        end
```
i would go the same way here as you did in the sessions controller: match in the method signature on whether the email param is there or not. we should put email first though, i think.

manukall commented on lib/phoenix_token_auth/registrator.ex:
```diff
@@ -8,13 +8,23 @@ defmodule PhoenixTokenAuth.Registrator do
+      |> UserHelper.validator
+      |> set_hashed_password
+    end
+  end
```
i would prefer allowing email and username attributes to be set at the same time here. (`~(username)` -> `~(email username)`)
then we need to run both uniqueness validations though.
then i would prefer to have the `changeset` method do the things both cases have in common and later call a private method that implements the logic that depends on whether we are in the username or email case. this method would match on the params again.

manukall commented:
oh tests for the new functionality would also be greatly appreciated.

Commit:
doing note: please rename to authenticate_by_email / authenticate_by_username

Commit:
removal of grant_type as asked in note

Commit:
match the mailer signature to check if email is present

Commit:
reworked registering function and created changeset_helper

add username to User model in tests

alexjp commented:
tried to do all your notes.
did not manage to make (~(username) -> ~(email username)) set at the same time, but made it better than i had.

if everything is ok, tests are coming :)

manukall commented on lib/phoenix_token_auth/registrator.ex:
```diff
@@ -7,15 +7,28 @@ defmodule PhoenixTokenAuth.Registrator do
+  end
+
+  def changeset_helper(changeset) do
+    changeset
```
was the problem with allowing both email and username that the uniqueness validation failed, if one of them is nil? if so, i think we should build a custom validator that only adds the error if the field is not nil. what do you think?

alexjp replied:
agree.

Commit:
added tests for username login and registration

last two login tests are not working ( due to map vs struct function
signature )

alexjp commented:
Added some tests.
I am having trouble with my "elixir", The registrator signature is expecting a map, but the login test is sending structs. That fails the last two login tests.

What is the best way around this ?

manukall commented:
that should actually work, as structs are maps. are the keys different? maybe atoms instead of strings?

Commit:
fixed the login tests and moved changeset_helper in registrator as private: defp

alexjp commented:
fixed the tests :)

manukall commented on lib/phoenix_token_auth/mailer.ex:
```diff
@@ -25,6 +25,7 @@ defmodule PhoenixTokenAuth.Mailer do
   Both config fields have to be functions returning binaries. welcome_email_subject receives the user and
   welcome_email_body the user and confirmation token.
   """
+  def send_welcome_email(user = %{email: email}, confirmation_token) when email == nil do end
```
This is my last comment, after that I will merge. Promise!
This line throws a warning, because confirmation_token isn't used. I think we can just remove the whole line anyway, right?

alexjp replied:
No worries... the more comments, the more i fix things :)

Its because of this code on users_controller:

```
if changeset.valid? do
      case Util.repo.transaction fn ->
        user = Util.repo.insert(changeset)
        Mailer.send_welcome_email(user, confirmation_token)
      end do
        {:ok, _} -> json conn, :ok
      end
    else
      Util.send_error(conn, Enum.into(changeset.errors, %{}))
    end
```

So, it calls the mailer.send_welcome_email, but if email is nil, it just does nothing.

manukall replied:
oh i thought we changed that in the users_controller and moved to two function bodies. i think i would actually like the approach with the two function bodies in users_controller better. the main reason is, that it's more explicit that way and `send_welcome_email` still raises an error when the user has no email, because of a bug or something,
would you change that or are you too annoyed already? :wink:

alexjp replied:
i made it simpler, just a simple conditional if user has email should do.
not really annoyed, but i m trying to make my mindset change from OOP to FP, i thought that the previous way was the correct one ( use functions and match parameters )

Commit:
remove warnings of unused variables in mailer

Commit:
remove empty send_welcome_email function

put sending email in registration based condition if user has email
remove test that only tested the empty send_welcome_email function

Commit:
added the id of the user record into the token response

this should make it much easier to fetch the correct "account" resourse

alexjp commented:
Added the user id into the session response.
This should make it alot easier to get the user id.

Previously, it needed to:
- do login
- do request to account
- do request to proper user model with query based on email or username

now should be:
- do login
- do request to proper user model, fetching the proper resource by id

The account is nice, but there is two problems for me:
- not big deal, but would have to create an account model just to store the username and email
- ( very important ) would have to format the account response according to jsonapi for ember

manukall commented:
Review status: 0 of 11 files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

<sup>**[lib/phoenix_token_auth/mailer.ex, line 0 [r4]](https://reviewable.io:443/reviews/manukall/phoenix_token_auth/21#-JuLHXzGdr-uujbtWPOg-r4-28)** ([raw file](https://github.com/manukall/phoenix_token_auth/blob/a70ad9fdf9dde82ff247241f80b6a9151ccbbd04/lib/phoenix_token_auth/mailer.ex#L0)):</sup>
i think using functions instead of conditions is indeed preferred in elixir. that's why i would suggest the approach of using two implementations of UserController#create. I imagined something like:

``` elixir
  def create(conn, %{"user" => %{email: email}}) when email != ""... do
      {confirmation_token, changeset} = Registrator.changeset(params)
      |> Confirmator.confirmation_needed_changeset

    if changeset.valid? do
      case Util.repo.transaction fn ->
        user = Util.repo.insert(changeset)
             Mailer.send_welcome_email(user, confirmation_token)
      end do
        {:ok, _} -> json conn, :ok
      end
    else
      Util.send_error(conn, Enum.into(changeset.errors, %{}))
    end

    def create(conn, %{"user" => params}) do
        changeset = Registrator.changeset(params)

    if changeset.valid? do
      case Util.repo.transaction fn ->
        user = Util.repo.insert(changeset)
        end do
        {:ok, _} -> json conn, :ok
      end
    else
      Util.send_error(conn, Enum.into(changeset.errors, %{}))
    end
  end
```

( note: this is only badly formatted pseudo code)

the reason i didn't like the two send_email method implmementations was just that i think an error should be raised at that point if no email is present. in the controller we know that there are two kinds of creating a user and handle that: one with email and one with password. we should try to keep that logic somewhat local there, so we do not accidently swallow errors in the mailers that are real errors.

---

Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/manukall/phoenix_token_auth/21)

<!-- Sent from Reviewable.io -->

alexjp commented:
Understood... i tried doing something like that but saw that i was duplicating alot of code just because an if.

I will update the code accordingly. thanks !

Commit:
update create functions according to note

Commit:
* update ecto version

manukall commented:
hey @alexjp,
what's the state here? there is one failing test and lot's of deprecation warnings...

alexjp commented:
Sorry, it was because i updated ecto version, and insert/2 and update/2 are now insert!/2 and update!/2.

I have been away from computer there last few days, but will update it accordingly.

Commit:
update to latest ecto, fixing ecto warnings

fix tests that got broken with function guards of username/email

Commit:
include bad removed tests. made email default in registrator again (
removed function guard for email )

alexjp commented:
@manukall very lots of sorry for the delay, should be all fixed.

manukall commented:
@alexjp thanks again! great work!!! i'll merge the refactor branch into master and release a new version with your changes now...

alexjp commented:
really, thank you @manukall , for phoenix_token_auth and the help in this pull request.

I will try to post online in the weekend, an example of server api in phoenix and web app in emberjs. It could be included in the readme, for help references.

The app actually is already done, only missing removing my repository and moving it to this one.

With this and jsonapi, i figure elixir and phoenix opens alot of doors for emberjs developers!!!

thanks !!!!!!
